python/lsst/daf/persistence/butler.py

```diff
@@ -136,44 +190,165 @@ def __init__(self, root, mapper=None, **mapperArgs):
-            config = Butler.cfg(repoCfg=repoCfg)
-        self._initWithCfg(config)
-
-    def _initWithCfg(self, cfg):
-        """Initialize this Butler with cfg
+        self._initArgs = {'root':root, 'mapper':mapper, 'inputs':inputs, 'outputs':outputs, 
+                          'mapperArgs':mapperArgs}
+
+        isLegacyRepository = inputs is None and outputs is None
+
+        if root is not None and not isLegacyRepository:
+            raise RuntimeError(
+                'root is a deprecated parameter and may not be used with the parameters input and output')
+
+        if isLegacyRepository:
+            if root is None:
+                if hasattr(mapper, 'root'):
+                    # in legacy repos, the mapper may be given the root directly.
+                    root = mapper.root
+                else:
+                    # in the past root="None" could be used to mean root='.'
+                    root = '.'
+            outputs = RepositoryArgs(mode='rw',
+                                     root=root, 
+                                     mapper=mapper, 
+                                     mapperArgs=mapperArgs)
+            outputs.isLegacyRepository = True
 
-        :param cfg: a dict (or daf_persistence.Policy) of key-value pairs that describe the configuration.
-                    Best practice is to create this object by calling Butler.cfg()
-        :return: None
-        """
-        self._cfg = cfg
         self.datasetTypeAliasDict = {}
 
-        self.repository = Repository.makeFromCfg(self._cfg['repoCfg'])
-
         # Always use an empty Persistence policy until we can get rid of it
         persistencePolicy = pexPolicy.Policy()
         self.persistence = Persistence.getPersistence(persistencePolicy)
         self.log = pexLog.Log(pexLog.Log.getDefaultLog(), "daf.persistence.butler")
+        
+        inputs = listify(inputs)
+        outputs = listify(outputs)
+
+        # if only a string is passed for inputs or outputs, assumption is that it's a URI;
+        # place it in a RepositoryArgs instance; cfgRoot for inputs, root for outputs.
+        for n in xrange(len(inputs)):
+            if isinstance(inputs[n], basestring):
+                inputs[n] = RepositoryArgs(cfgRoot=inputs[n])
+        for n in xrange(len(outputs)):
+            if isinstance(outputs[n], basestring):
+                outputs[n] = RepositoryArgs(root=outputs[n])
+
+        self._repos = RepoDataContainer()
+
+        defaultMapper = self.getDefaultMapper(inputs)
+
+        for args in outputs:
+            self._addRepo(args, inout='out', defaultMapper=defaultMapper)
+
+        for args in inputs:
+            self._addRepo(args, inout='in')
+
+        for repoData in self._repos.outputs():
+            inputCfgRoots = [rd.args.cfgRoot for rd in self._repos.inputs() if rd is not repoData]
+            repoData.cfg.addParents(inputCfgRoots)
+            Storage.putRepositoryCfg(repoData.cfg, repoData.args.cfgRoot)
```

Comment:
This overwrites the existing configuration with the new parents, correct?  A comment to that effect might be good.  I'm also worried about how all of these `put`s are going to work when run in parallel across many nodes.  Please make sure that things like `matchesArgs()` will work properly if some nodes see the parents added and some do not.